pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -346,136 +535,226 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *api.Ser
-		VipSubnetID: lbaas.opts.SubnetId,
-	}
 
-	loadBalancerIP := apiService.Spec.LoadBalancerIP
-	if loadBalancerIP != "" {
-		createOpts.VipAddress = loadBalancerIP
-	}
-
-	loadbalancer, err := loadbalancers.Create(lbaas.network, createOpts).Extract()
+	oldListeners, err := getListenersByLoadBalancerID(lbaas.network, loadbalancer.ID)
 	if err != nil {
-		// cleanup what was created so far
-		_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
-		return nil, err
+		return nil, fmt.Errorf("Error getting LB %s listeners: %v", name, err)
 	}
-
-	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
-
 	for portIndex, port := range ports {
-		listener, err := listeners.Create(lbaas.network, listeners.CreateOpts{
-			Name:           fmt.Sprintf("listener_%s_%d", name, portIndex),
-			Protocol:       listeners.Protocol(port.Protocol),
-			ProtocolPort:   int(port.Port),
-			LoadbalancerID: loadbalancer.ID,
-		}).Extract()
-		if err != nil {
-			// cleanup what was created so far
-			_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
-			return nil, err
+		listener := getListenerForPort(oldListeners, port)
+		if listener == nil {
+			glog.V(4).Infof("Creating listener for port %d", int(port.Port))
+			listener, err = listeners.Create(lbaas.network, listeners.CreateOpts{
+				Name:           fmt.Sprintf("listener_%s_%d", name, portIndex),
+				Protocol:       listeners.Protocol(port.Protocol),
+				ProtocolPort:   int(port.Port),
+				LoadbalancerID: loadbalancer.ID,
+			}).Extract()
+			if err != nil {
+				// Unknown error, retry later
+				return nil, fmt.Errorf("Error creating LB listener: %v", err)
+			}
+			waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
 		}
 
-		waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+		glog.V(4).Infof("Listener for %s port %d: %s", string(port.Protocol), int(port.Port), listener.ID)
 
-		pool, err := v2_pools.Create(lbaas.network, v2_pools.CreateOpts{
-			Name:        fmt.Sprintf("pool_%s_%d", name, portIndex),
-			Protocol:    v2_pools.Protocol(port.Protocol),
-			LBMethod:    lbmethod,
-			ListenerID:  listener.ID,
-			Persistence: persistence,
-		}).Extract()
-		if err != nil {
-			// cleanup what was created so far
-			_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
-			return nil, err
+		// After all ports have been processed, remaining listeners are removed as obsolete.
+		// Pop valid listeners.
+		oldListeners = popListener(oldListeners, listener.ID)
+		pool, err := getPoolByListenerID(lbaas.network, loadbalancer.ID, listener.ID)
+		if err != nil && err != ErrNotFound {
+			// Unknown error, retry later
+			return nil, fmt.Errorf("Error getting pool for listener %s: %v", listener.ID, err)
+		}
+		if pool == nil {
+			glog.V(4).Infof("Creating pool for listener %s", listener.ID)
+			pool, err = v2_pools.Create(lbaas.network, v2_pools.CreateOpts{
+				Name:        fmt.Sprintf("pool_%qs_%d", name, portIndex),
```

Comment:
This results in some pretty odd looking names :)

	$ neutron lbaas-pool-list
	+--------------------------------------+--------------------------------------------+----------+----------------+
	| id                                   | name                                       | protocol | admin_state_up |
	+--------------------------------------+--------------------------------------------+----------+----------------+
	| 6f5f4cf6-a6c9-41d9-831f-98cde0a972f6 | pool_"ab8200e11858e11e687f7fa163e76dca"s_0 | TCP      | True           |
	| dd187fbe-059d-4cd4-bc8c-d68c2bad5bfc | pool_"ab5f92815858e11e687f7fa163e76dca"s_0 | TCP      | True           |
	+--------------------------------------+--------------------------------------------+----------+----------------+